---
title: "1. Re-analysis Transgen - Impact of sequencing depth"
author: "adapted from Branko Cirovic"
date: "20 10 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Load Libraries
```{r, message=FALSE, warning=FALSE}
library(tximport)
library(readr)
library(DESeq2)
library(dplyr)
library(biomaRt)
library(AnnotationDbi)
library(BiocParallel)
library(ggplot2)
library(RColorBrewer)
library(plotly)
library(pheatmap)
library(ggpubr)
library(fgsea)
library(dplyr)
library(tidyr)
library(tibble)
library(ggbeeswarm)
library(RColorBrewer)
library(patchwork)
library(ggrepel)
```

## Functions

```{r}
plotMA <- function(DEresults,
                   ylim=c(-2,2),
                   padjThreshold=0.05,
                   xlab = "baseMean",
                   ylab = expression(log[2]~fold~change),
                   log = "x",
                   cex=0.45){
  x <- as.data.frame(DEresults)
  if (!(is.data.frame(x) && all(c("baseMean", "log2FoldChange") %in% colnames(x)))){
    stop("'x' must be a data frame with columns named 'baseMean', 'log2FoldChange'.")
  }
  col = ifelse(x$padj>=padjThreshold, "gray32", "red3")
  cex =  ifelse(x$padj>=padjThreshold, 0.3, 0.8)
  py = x$log2FoldChange
  if(missing(ylim)){
    ylim = c(-1,1) * quantile(abs(py[is.finite(py)]), probs=0.99) * 1.1
  }
  plot(x=x$baseMean,
       y=pmax(ylim[1], pmin(ylim[2], py)),
       log=log,
       pch=ifelse(py<ylim[1], 6, ifelse(py>ylim[2], 2, 16)),
       cex=cex,
       col=col,
       size=3,
       xlab=xlab,
       ylab=ylab,
       ylim=ylim,
       main="MA plot")
  abline(h=0, lwd=4, col="#ff000080")
  abline(h=c(-1.5,1.5), lwd=2, col="dodgerblue")
}


plotVolcano <-  function(DEresults,
                         labelnum=20){
  
  # specify labeling
  DEresults <- as.data.frame(DEresults)
  upDE <-  as.data.frame(DEresults[DEresults$regulation =="up", ])
  FClabel_up <- upDE[order(abs(upDE$log2FoldChange), decreasing = TRUE),]
  if(nrow(FClabel_up)>labelnum){
    FClabel_up <- as.character(FClabel_up[c(1:labelnum),"GENEID"])
  } else {
    FClabel_up <- as.character(FClabel_up$GENEID)}
  plabel_up <- upDE[order(upDE$padj, decreasing = FALSE),]
  if(nrow(plabel_up)>labelnum){
    plabel_up <- as.character(plabel_up[c(1:labelnum),"GENEID"])
  } else {
    plabel_up <- as.character(plabel_up$GENEID)}
  
  downDE <-  as.data.frame(DEresults[DEresults$regulation =="down",])
  FClabel_down <- downDE[order(abs(downDE$log2FoldChange), decreasing = TRUE),]
  if(nrow(FClabel_down)>labelnum){
    FClabel_down <- as.character(FClabel_down[c(1:labelnum),"GENEID"])
  } else {
    FClabel_down <- as.character(FClabel_down$GENEID)}
  plabel_down <- downDE[order(downDE$padj, decreasing = FALSE),]
  if(nrow(plabel_down)>labelnum){
    plabel_down <- as.character(plabel_down[c(1:labelnum),"GENEID"])
  } else {
    plabel_down <- as.character(plabel_down$GENEID)}
  
  
  label<- unique(c(FClabel_up, plabel_up, FClabel_down, plabel_down))
  
  data <- as.data.frame(DEresults)
  data$label<- ifelse(data$GENEID %in% label == "TRUE", as.character(data$GENEID), "")
  data <- data[,colnames(data) %in% c("label", "log2FoldChange", "padj", "regulation")]
  
  # Volcano Plot
  ggplot(data=na.omit(data), aes(x=log2FoldChange, y=-log10(padj), colour=regulation)) +
    geom_point(alpha=0.4, size=1.75) +
    scale_color_manual(values=c(down = "cornflowerblue", "n.s." = "grey", "up" = "firebrick"))+
    scale_x_continuous() +
    scale_y_continuous() +
    xlab("log2(FoldChange)") +
    ylab("-log10(padj)") +
    geom_vline(xintercept = 0, colour="black")+
    geom_vline(xintercept = c(-log(1.5,2),log(1.5,2)), colour="red")+
    geom_hline(yintercept=-log(0.05,10),colour="red")+
    geom_text_repel(data=na.omit(data[!data$label =="",]),aes(label=label), size=3)+
    guides(colour="none") +
    ggtitle("Volcano Plot") +
    theme_bw()
}

plotVolcano_adj <-  function(DEresults,
                         celltype = c("Mono", "cMOP"),
                         labelnum=20){
  
  # specify labeling
  DEresults <- as.data.frame(DEresults)
  upDE <-  as.data.frame(DEresults[DEresults$regulation =="up", ])
  downDE <-  as.data.frame(DEresults[DEresults$regulation =="down",])
  
  if (celltype == "Mono"){
    label<- unique(c(DEGs_publication$Mono_down, DEGs_publication$Mono_up))
  } else if (celltype == "cMOP") {
    label<- unique(c(DEGs_publication$CMOP_down, DEGs_publication$CMOP_up))
  }
  
  data <- as.data.frame(DEresults)
  data$label<- ifelse(data$GENEID %in% label == "TRUE", as.character(data$GENEID), "")
  data <- data[,colnames(data) %in% c("label", "log2FoldChange", "padj", "regulation")]
  
  # Volcano Plot
  ggplot(data=na.omit(data), aes(x=log2FoldChange, y=-log10(padj), colour=regulation)) +
    geom_point(alpha=0.4, size=1.75) +
    geom_point(data=na.omit(data[!data$label =="",]), aes(fill = regulation), alpha=0.7, size=3, pch = 21, color = "black") +
    scale_color_manual(values=c(down = "cornflowerblue", "n.s." = "grey", "up" = "firebrick"))+
    scale_fill_manual(values=c(down = "cornflowerblue", "n.s." = "grey", "up" = "firebrick"))+
    scale_x_continuous() +
    scale_y_continuous() +
    xlab("log2(FoldChange)") +
    ylab("-log10(padj)") +
    geom_vline(xintercept = 0, colour="black")+
    geom_vline(xintercept = c(-log(1.5,2),log(1.5,2)), colour="red")+
    geom_hline(yintercept=-log(0.05,10),colour="red")+
    geom_text_repel(data=na.omit(data[!data$label =="",]),aes(label=label), size=3)+
    guides(colour="none") +
    ggtitle("Volcano Plot") +
    theme_bw()
}

```


# Import 

### Published DEGs
```{r}
DEGs_publication <- read.csv(file = "DEGs_publication.csv", sep = ";")
```


### sample table
```{r}
sample_table <- read.csv2("sample_table.csv", sep=",", row.names = 1)
```


### Transcript/Gene Annotation

```{r}
# tx_annotation <- read.delim(file = "~/data/annotation/mouse/mm10/ID2SYMBOL_gencode_vM16_transcript.txt", 
#                             header = F , 
#                             stringsAsFactors = F, 
#                             col.names = c("GENEID", "TXNAME", "SYMBOL", "GENETYPE"))
```


### Import counts

```{r}
# Otximp <- tximport(files, type = "kallisto", tx2gene = tx_annotation[,2:3]) 
```

# Build DESeqDataSet
```{r}
# Odds_TG <- DESeqDataSetFromTximport(Otximp, sample_table, design=~Condition)

Odds_TG <- readRDS("Odds_TG.rds")
Odds_TG
```

# Filter DESeq Object

### Pre-filter counts
```{r}
Okeep_new <- rowSums(counts(Odds_TG)) >= 10
Odds_TG_f <- Odds_TG[Okeep_new,]
Odds_TG_f
```


### Filter Genes

#### Gm PseudoGenes
```{r}
#this will count all Gene names except that start with an Gm+digit = pseudogenes 
length(grep("Gm\\d",rownames(Odds_TG_f),perl = T,invert = T))
```

```{r}
noGMgenes=(grep("Gm\\d",rownames(Odds_TG_f),perl = T,invert = T))
```

Subset to remove Gm pseudogenes 
```{r}
Odds_TG_f=Odds_TG_f[noGMgenes,]
```

#### Rik Genes
```{r}
#this will count all Gene names except that end with Rik or Rik#
length(grep("Rik$|Rik\\d$",rownames(Odds_TG_f),perl = T,invert = T))
```


```{r}
#this will call all Gene names except that end with Rik
noRikgenes=(grep("Rik$|Rik\\d$",rownames(Odds_TG_f),perl = T,invert = T))
```

Subset to remove Rik Genes
```{r}
Odds_TG_f=Odds_TG_f[noRikgenes,]
dim(Odds_TG_f)
```

## Reformat
```{r}
Odds_TG_f_1 <- Odds_TG_f
```

```{r}
Odds_TG_f_1 <- Odds_TG_f_1[sort(rownames(Odds_TG_f_1)),]
```


## Normalization
```{r}
# estimate normalization factors
Odds_TG_d <- DESeq(Odds_TG_f_1)
```

```{r}
norm <- counts(Odds_TG_d, normalized=T)
```



# Differential Expression Analysis

List Conditions
"GMP_control"   
"GMP_infected"  
"Ly6Ch_control" 
"Ly6Ch_infected"
"cMOP_control"  
"cMOP_infected"


## Comparison GMP_control vs GMP_infected
```{r}
res_GMP <- results(Odds_TG_d,contrast=c("Condition", "GMP_infected", "GMP_control"),parallel = T)

res_GMP$regulation <- ifelse(!is.na(res_GMP$padj) & res_GMP$padj < 0.05 & res_GMP$log2FoldChange > 0, "up",
                             ifelse(!is.na(res_GMP$padj) & res_GMP$padj < 0.05 & res_GMP$log2FoldChange < 0, "down", "n.s."))
res_GMP$GENEID <- rownames(res_GMP)
```

Filter and Order Results for res_GMP
```{r}
## subset data
res_GMP_Sig <- subset(res_GMP, padj < 0.05)
## Order this list according to fold change
res_GMP_Sig=res_GMP_Sig[order(res_GMP_Sig$log2FoldChange, decreasing=T), ]
```


### Number DEGs
```{r}
df <- as.data.frame(res_GMP)
df$regFC <- ifelse(!is.na(df$padj) & df$padj < 0.05 & df$log2FoldChange > log(1.5,2), "up",
                             ifelse(!is.na(df$padj) & df$padj < 0.05 & df$log2FoldChange < -log(1.5,2), "down", "n.s."))
table(df$regFC)
table(df$regulation)
```
```{r, fig.height=4, fig.width=8}
tmp <- df %>% subset(regulation != "n.s.") %>% group_by(regulation) %>% summarize(n = n())

p1 <- ggplot(tmp, aes(x=regulation, y = n, fill = regulation)) + 
  geom_bar(stat = "identity", position="dodge") + 
  geom_text(aes(label = n), size = 4, position = position_dodge(0.9), vjust = -0.2) +
  ylim(0, 5) + 
  scale_fill_manual(values= c("dodgerblue", "firebrick")) + 
  ggtitle("Number of DEGs (padj < 0.05)") + xlab("") + 
  theme_bw() 

tmp <- df %>% subset(regFC != "n.s.") %>% group_by(regFC) %>% summarize(n = n())

p2 <- ggplot(tmp, aes(x=regFC, y = n, fill = regFC)) + 
  geom_bar(stat = "identity", position="dodge") +
  geom_text(aes(label = n), size = 4, position = position_dodge(0.9), vjust = -0.2) +
  ylim(0, 5) +
  scale_fill_manual(values= c("dodgerblue", "firebrick")) + 
  ggtitle("Number of DEGs (padj < 0.05, FC cutoff 1.5)") + 
  theme_bw() +xlab("") + 
  theme(legend.position = "none")

p1 + p2 + plot_layout(guides = "collect")
```


### MA plot
```{r}
plotMA(DEresults = res_GMP)
```
### Volcano plot
```{r, fig.height=4, fig.width=5}
plotVolcano(DEresults = res_GMP)
```


### Barplots 

```{r, fig.height=4, fig.width=4}
df <- norm[rownames(norm) %in% rownames(res_GMP_Sig), grepl(pattern="GMP", x = colnames(norm))] %>% as.data.frame() %>% rownames_to_column("gene")
df_long <- df %>% pivot_longer(!gene, names_to = "ID", values_to = "value")

up <- rownames(res_GMP_Sig[res_GMP_Sig$log2FoldChange>0,])
down <- rownames(res_GMP_Sig[res_GMP_Sig$log2FoldChange<0,])
df_long$gene <- factor(df_long$gene, levels = c(up, down))

ggplot(df_long, aes(x=gene, y=value)) + 
  geom_boxplot(outlier.colour = NA) +
  geom_jitter(aes(color=ID)) + 
  scale_color_manual(values= c(brewer.pal(name = "Oranges", n = 9)[4:6], brewer.pal(name = "Purples", n = 9)[4:6])) + 
  ggtitle("Expression levels of DEGs") + ylab("Normalized counts (mean)") +
  theme_bw()+
  theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1), legend.position = "bottom") 
```


### Heatmaps
```{r, fig.height=3, fig.width=5}
pheatmap(norm[rownames(norm) %in% rownames(res_GMP_Sig), grepl(pattern = "GMP", x = colnames(norm))],
         scale = "none")
pheatmap(norm[rownames(norm) %in% rownames(res_GMP_Sig), grepl(pattern = "GMP", x = colnames(norm))],
         scale = "row")
```


## Comparison Ly6Ch_control vs Ly6Ch_infected
```{r}
res_Mono <- results(Odds_TG_d,contrast=c("Condition", "Ly6Ch_infected", "Ly6Ch_control"),parallel = T)

res_Mono$regulation <- ifelse(!is.na(res_Mono$padj) & res_Mono$padj < 0.05 & res_Mono$log2FoldChange > 0, "up",
                             ifelse(!is.na(res_Mono$padj) & res_Mono$padj < 0.05 & res_Mono$log2FoldChange < 0, "down", "n.s."))
res_Mono$GENEID <- rownames(res_Mono)
```

Filter and Order Results for res_Mono
```{r}
## subset data
res_Mono_Sig <- subset(res_Mono, padj < 0.05)
## Order this list according to fold change
res_Mono_Sig=res_Mono_Sig[order(res_Mono_Sig$log2FoldChange, decreasing=T), ]
```

### Number DEGs
```{r}
df <- as.data.frame(res_Mono)
df$regFC <- ifelse(!is.na(df$padj) & df$padj < 0.05 & df$log2FoldChange > log(1.5,2), "up",
                             ifelse(!is.na(df$padj) & df$padj < 0.05 & df$log2FoldChange < -log(1.5,2), "down", "n.s."))
table(df$regFC)
table(df$regulation)
```
```{r, fig.height=4, fig.width=8}
tmp <- df %>% subset(regulation != "n.s.") %>% group_by(regulation) %>% summarize(n = n())

p1 <- ggplot(tmp, aes(x=regulation, y = n, fill = regulation)) + 
  geom_bar(stat = "identity", position="dodge") + 
  geom_text(aes(label = n), size = 4, position = position_dodge(0.9), vjust = -0.2) +
  ylim(0, 70) + 
  scale_fill_manual(values= c("dodgerblue", "firebrick")) + 
  ggtitle("Number of DEGs (padj < 0.05)") + xlab("") + 
  theme_bw() 

tmp <- df %>% subset(regFC != "n.s.") %>% group_by(regFC) %>% summarize(n = n())

p2 <- ggplot(tmp, aes(x=regFC, y = n, fill = regFC)) + 
  geom_bar(stat = "identity", position="dodge") +
  geom_text(aes(label = n), size = 4, position = position_dodge(0.9), vjust = -0.2) +
  ylim(0, 70) +
  scale_fill_manual(values= c("dodgerblue", "firebrick")) + 
  ggtitle("Number of DEGs (padj < 0.05, FC cutoff 1.5)") + 
  theme_bw() +xlab("") + 
  theme(legend.position = "none")

p1 + p2 + plot_layout(guides = "collect")
```

### MA plot
```{r}
plotMA(DEresults = res_Mono)
```


### Volcano plot
```{r, fig.height=6, fig.width=6}
plotVolcano(DEresults = res_Mono,labelnum = 5)
```


### Barplots

```{r, fig.height=4, fig.width=12}
df <- norm[rownames(norm) %in% rownames(res_Mono_Sig), grepl(pattern = "Ly6Ch", x = colnames(norm))] %>% as.data.frame() %>% rownames_to_column("gene")
df_long <- df %>% pivot_longer(!gene, names_to = "ID", values_to = "value")

up <- rownames(res_Mono_Sig[res_Mono_Sig$log2FoldChange>0,])
down <- rownames(res_Mono_Sig[res_Mono_Sig$log2FoldChange<0,])
df_long$gene <- factor(df_long$gene, levels = c(up, down))

ggplot(df_long, aes(x=gene, y=value)) + 
  geom_boxplot(outlier.color = NA) +
  geom_jitter(aes(color=ID)) + 
  scale_color_manual(values= c(brewer.pal(name = "Oranges", n = 9)[3:7], brewer.pal(name = "Purples", n = 9)[3:7])) + 
  ggtitle("Expression level of DEGs") + ylab("Normalized counts (mean)") +
  theme_bw() + 
  theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1), legend.position = "bottom") 

```


### Heatmap

```{r, fig.height=15, fig.width=5}
pheatmap(norm[rownames(norm) %in% rownames(res_Mono_Sig), grepl(pattern = "Ly6Ch", x = colnames(norm))], 
         scale = "none")
pheatmap(norm[rownames(norm) %in% rownames(res_Mono_Sig), grepl(pattern = "Ly6Ch", x = colnames(norm))], 
         scale = "row")
```

## Comparison cMOP_control vs cMOP_infected
```{r}
res_cMOP <- results(Odds_TG_d, contrast=c("Condition", "cMOP_infected","cMOP_control"),parallel = T)

res_cMOP$regulation <- ifelse(!is.na(res_cMOP$padj) & res_cMOP$padj < 0.05 & res_cMOP$log2FoldChange > 0, "up",
                             ifelse(!is.na(res_cMOP$padj) & res_cMOP$padj < 0.05 & res_cMOP$log2FoldChange < 0, "down", "n.s."))
res_cMOP$GENEID <- rownames(res_cMOP)
```

Filter and Order Results for cMOP_infected
```{r}
## subset data
res_cMOP_Sig <- subset(res_cMOP, padj < 0.05)
## Order this list according to fold change
res_cMOP_Sig=res_cMOP_Sig[order(res_cMOP_Sig$log2FoldChange, decreasing=T), ]
```


### Number DEGs
```{r}
df <- as.data.frame(res_cMOP)
df$regFC <- ifelse(!is.na(df$padj) & df$padj < 0.05 & df$log2FoldChange > log(1.5,2), "up",
                             ifelse(!is.na(df$padj) & df$padj < 0.05 & df$log2FoldChange < -log(1.5,2), "down", "n.s."))
table(df$regFC)
table(df$regulation)
```

```{r, fig.height=4, fig.width=8}
tmp <- df %>% subset(regulation != "n.s.") %>% group_by(regulation) %>% summarize(n = n())

p1 <- ggplot(tmp, aes(x=regulation, y = n, fill = regulation)) + 
  geom_bar(stat = "identity", position="dodge") + 
  geom_text(aes(label = n), size = 4, position = position_dodge(0.9), vjust = -0.2) +
  ylim(0, 35) + 
  scale_fill_manual(values= c("dodgerblue", "firebrick")) + 
  ggtitle("Number of DEGs (padj < 0.05)") + xlab("") + 
  theme_bw() 

tmp <- df %>% subset(regFC != "n.s.") %>% group_by(regFC) %>% summarize(n = n())

p2 <- ggplot(tmp, aes(x=regFC, y = n, fill = regFC)) + 
  geom_bar(stat = "identity", position="dodge") +
  geom_text(aes(label = n), size = 4, position = position_dodge(0.9), vjust = -0.2) +
  ylim(0, 35) +
  scale_fill_manual(values= c("dodgerblue", "firebrick")) + 
  ggtitle("Number of DEGs (padj < 0.05, FC cutoff 1.5)") + 
  theme_bw() +xlab("") + 
  theme(legend.position = "none")

p1 + p2 + plot_layout(guides = "collect")
```

### MA plot
```{r}
plotMA(DEresults = res_cMOP)
```


### Volcano plot
```{r, fig.height=5, fig.width=7}
plotVolcano(DEresults = res_cMOP,labelnum = 5)
```


### Barplots

```{r, fig.height=4, fig.width=8}
df <- norm[rownames(norm) %in% rownames(res_cMOP_Sig), grepl(pattern = "cMOP", x = colnames(norm))] %>% as.data.frame() %>% rownames_to_column("gene")
df
df_long <- df %>% pivot_longer(!gene, names_to = "ID", values_to = "value")
df_long

up <- rownames(res_cMOP_Sig[res_cMOP_Sig$log2FoldChange>0,])
down <- rownames(res_cMOP_Sig[res_cMOP_Sig$log2FoldChange<0,])
df_long$gene <- factor(df_long$gene, levels = c(up, down))

ggplot(df_long, aes(x=gene, y=value)) + 
  geom_boxplot() +
  geom_jitter(aes(color=ID)) + 
  scale_color_manual(values= c(brewer.pal(name = "Oranges", n = 9)[3:7], brewer.pal(name = "Purples", n = 9)[3:7])) + 
  ggtitle("Expression level of DEGs") + ylab("Normalized counts") +
  theme_bw()+ 
  theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1), legend.position = "bottom") 
```


### Heatmaps
```{r, fig.height=10, fig.width=5}
pheatmap(norm[rownames(norm) %in% rownames(res_cMOP_Sig), grepl(pattern = "cMOP", x = colnames(norm))], 
         scale = "none")
pheatmap(norm[rownames(norm) %in% rownames(res_cMOP_Sig), grepl(pattern = "cMOP", x = colnames(norm))], 
         scale = "row")
```

## Overlapping DEGs

```{r}
table(rownames(res_Mono_Sig) %in% c(DEGs_publication$Mono_down, DEGs_publication$Mono_up))
table(rownames(res_cMOP_Sig) %in% c(DEGs_publication$CMOP_down, DEGs_publication$CMOP_up))
```

### Heatmaps - cMOPs
```{r, fig.height=4, fig.width=5}
genes <-  rownames(res_cMOP_Sig)[rownames(res_cMOP_Sig) %in% c(DEGs_publication$CMOP_down, DEGs_publication$CMOP_up)]
pheatmap(norm[rownames(norm) %in% genes, grepl(pattern = "cMOP", x = colnames(norm))], scale = "none")
pheatmap(norm[rownames(norm) %in% genes, grepl(pattern = "cMOP", x = colnames(norm))], scale = "row")
```


### Heatmaps  - Monos
```{r, fig.height=10, fig.width=5}
genes <-  rownames(res_Mono_Sig)[rownames(res_Mono_Sig) %in% c(DEGs_publication$Mono_down, DEGs_publication$Mono_up)]
pheatmap(norm[rownames(norm) %in% genes, grepl(pattern = "Ly6Ch", x = colnames(norm))], scale = "none")
pheatmap(norm[rownames(norm) %in% genes, grepl(pattern = "Ly6Ch", x = colnames(norm))], scale = "row")
```


## fgsea

### Reference files
GO BP full 
```{r}
# "c5.go.bp.v7.4.symbols.gmt" translated to mouse symbol and reformated
GO_bp_pathways_list <- readRDS("GO_bp_pathways_mouse.rds")
```

File used in original publication
```{r}
load("Myeloid_Inflamm_pathways_mouse.RData")
```

### Monocytes
```{r}
ranked.genes <- res_Mono %>% as.data.frame() %>% subset(!is.na(padj)) %>% arrange(desc(log2FoldChange)) %>% pull(log2FoldChange, GENEID)
fgsea_mono <- fgsea(pathways = GO_bp_pathways_list,
                    stats = ranked.genes,
                    minSize=15,
                    maxSize=700,
                    nperm=10000)

fgsea_mono_branko <- fgsea(pathways = inflamm_pathways_mouse,
                      stats = ranked.genes,
                      minSize=15,
                      maxSize=700,
                      nperm=10000)
```

Plot in normal RScript
```{r}
top3 <- fgsea_mono %>% arrange(pval) %>% slice_head(n=3)
plotGseaTable(GO_bp_pathways_list[top3$pathway], ranked.genes, fgsea_mono, 
              gseaParam=0.5)
```

```{r}
top3 <- fgsea_mono_branko %>% arrange(pval) %>% slice_head(n=3)
plotGseaTable(inflamm_pathways_mouse[top3$pathway], ranked.genes, fgsea_mono_branko, 
              gseaParam=0.5)
```



### Macrophages
```{r}
ranked.genes <- res_cMOP %>% as.data.frame() %>% subset(!is.na(padj)) %>% arrange(desc(log2FoldChange)) %>% pull(log2FoldChange, GENEID)
fgsea_cMOPs <- fgsea(pathways = GO_bp_pathways_list,
                    stats = ranked.genes,
                    minSize=15,
                    maxSize=700,
                    nperm=10000)

fgsea_cMOPs_branko <- fgsea(pathways = myeloid_pathways_mouse,
                      stats = ranked.genes,
                      minSize=15,
                      maxSize=700,
                      nperm=10000)
```

Plot in normal RScript
```{r}
top3 <- fgsea_cMOPs %>% arrange(pval) %>% slice_head(n=3)
plotGseaTable(GO_bp_pathways_list[top3$pathway], ranked.genes, fgsea_cMOPs, 
              gseaParam=0.5)
```


```{r}
top3 <- fgsea_cMOPs_branko %>% arrange(pval) %>% slice_head(n=3)
plotGseaTable(myeloid_pathways_mouse[top3$pathway], ranked.genes, fgsea_cMOPs_branko, 
              gseaParam=0.5)
```














